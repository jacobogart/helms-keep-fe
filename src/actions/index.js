export const storeQuests = quests => ({
  type: 'STORE_QUESTS',
  payload: {
    quests
  }
})

export const addQuest = quest => ({
  type: 'ADD_QUEST',
  payload: {
    quest
  }
})

export const editQuest = quest => {
  return {
    type: "EDIT_QUEST",
    payload: {
      quest
    }
  };
}

export const deleteQuest = id => {
  return {
    type: "DELETE_QUEST",
    payload: {
      id
    }
  };
}

export const showPopup = (bool, id, type) => {
  return {
    type: "SHOW_POPUP",
    payload: {
      bool,
      id,
      type
    }
  };
}

export const addExp = (exp) => {
  return {
    type: 'ADD_EXP',
    payload: {
      exp
    }
  }
}
export const addLvl = (lvl) => {
  return {
    type: 'ADD_LVL',
    payload: {
      lvl
    }
  }
}
export const addNextLvl = (lvl) => {
  return {
    type: 'ADD_NEXT_LVL',
    payload: {
      lvl
    }
  }
}