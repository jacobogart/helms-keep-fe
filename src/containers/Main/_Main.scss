.Main {
  margin-top: 155px;
  padding: 15px 25px;
}

.controls,
.headers {
  display: flex;
}

.page-title {
  color: white;
  font-weight: 500;
  font-size: 22px;
}

.view-all {
  margin-left: 15px;
  text-decoration: underline;
  cursor: pointer;
  font-size: 16px;
  margin-top: 5px;
  margin-right: -15px;
}

.main-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 20px;
  position: relative;
}

.search-input,
.quest-filter {
  height: 28px;
  margin-left: 30px;
}

.search-bar {
  position: relative;
}

.search-input {
  width: 325px;
  border-radius: 8px;
  border: none;
  padding-left: 10px;
}

.search-btn-container{
  position: absolute;
  width: 30px;
  height: 28px;
  right: 0px;
  color: white;
  border: none;
  border-radius: 0 8px 8px 0;
  &:hover {
    cursor: pointer;
  }
}

.quest-filter {
  width: 80px;
}

::placeholder {
  font-size: 14px;
  font-style: italic;
}

.add-quest-btn {
  position: absolute;
  display: flex;
  align-items: center;
  justify-content: center;
  text-decoration: none;
  bottom: 30px;
  right: 30px;
  height: 60px;
  width: 60px;
  border-radius: 100%;
  background: white;
  color: $btnHoverColor;
  font-size: 30px;
  box-shadow: 3px 3px 6px #000;

  &:hover {
    color: white;
    background: $btnColor;
    cursor: pointer;
  }
}

button:disabled {
  background: lightgrey!important;

  &:hover {
    background: lightgrey!important;
  }
}

