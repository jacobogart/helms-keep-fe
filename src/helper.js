export const searchQuests = (searchQuery, quests) => {
  let query = searchQuery.toLowerCase();
  let results = []; 
  quests.forEach(quest => {
    let challenges = quest.challenges.map(challenge => challenge.message.toLowerCase());
    if (quest.title.toLowerCase().includes(query) || challenges.some(challenge => challenge.includes(query))) {
      results.push(quest.id);
    }
  });
  return results
}