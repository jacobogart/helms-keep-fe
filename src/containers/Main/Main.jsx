import React, { Component } from "react";
import QuestContainer from "../QuestContainer/QuestContainer";
import { NavLink } from "react-router-dom";
import Dialog from "../Dialog/Dialog";
import { connect } from "react-redux";
import { searchQuests } from "../../helper";
import { addQuest } from "../../actions";
import PropTypes from "prop-types";
export class Main extends Component {
  state = {
    query: "",
    results: null
  };

  handleChange = e => {
    let query = e.target.value;
    if (!query) {
      this.setState({ results: null });
    }
    this.setState({ query });
  };

  clearResults = () => {
    this.setState({ results: null });
  };

  handleSubmit = e => {
    e.preventDefault();
    const results = searchQuests(this.state.query, this.props.quests);
    this.setState({ results });
  };

  render() {
    const { quests } = this.props;
    const { results } = this.state;
    const popup = this.props.popup.bool ? <Dialog /> : null;
    const viewAll =
      results && results.length !== quests.length ? (
        <p
          className="page-title view-all"
          role="button"
          onClick={this.clearResults}
        >
          View All Quests
        </p>
      ) : null;

    return (
      <main className="Main">
        {popup}
        <div className="main-header">
          <h1 className="page-title">My Quests</h1>
          <div className="controls">
            {viewAll}
            <form className="search-bar" onSubmit={this.handleSubmit}>
              <input
                className="search-input"
                placeholder="Search your quests..."
                onChange={this.handleChange}
              />
              <button type="submit" className="search-btn-container">
                <i className="fas fa-search" role="button" />
              </button>
            </form>
            {/* <select className="quest-filter" /> */}
          </div>
        </div>
        <QuestContainer results={this.state.results} />
        <NavLink to="/new-quest" className="add-quest-btn" type="button">
          <i className="fas fa-plus" />
        </NavLink>
      </main>
    );
  }
}

export const mapStateToProps = state => ({
  popup: state.popup,
  quests: state.quests
});

export const mapDispatchToProps = dispatch => ({
  addQuest: quest => dispatch(addQuest(quest))
});

Main.propTypes = {
  addQuest: PropTypes.func,
  popup: PropTypes.object,
  quests: PropTypes.arrayOf(PropTypes.object)
};
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Main);
