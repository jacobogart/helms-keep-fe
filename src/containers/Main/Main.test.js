import React from 'react';
import { shallow } from 'enzyme';
import { Main } from '../Main/Main';
import * as helper from '../../helper';
import { mapStateToProps, mapDispatchToProps } from './Main';
import { addQuest } from '../../actions';

describe('Main', () => {
  let wrapper;
  const mockQuest = [
    { id: 3, title: 'My first quest', challenges: ['hi', 'there'] }
  ];
	const mockQuests = [
		{ id: 1, title: 'My first quest', challenges: [ 'hi', 'there' ] },
		{ id: 2, title: 'My second quest', challenges: [ 'beef', 'cake' ] }
	];
	const mockAddQuest = jest.fn();
	const mockEvent = {
		target: { value: 'beef' },
		preventDefault: jest.fn()
	};

	helper.searchQuests = jest.fn().mockImplementation(() => mockQuests);

	beforeEach(() => {
		wrapper = shallow(<Main popup={{}} quests={mockQuests} addQuest={mockAddQuest} />);
	});

	it('should match the snapshot when the store prop popup is false', () => {
		expect(wrapper).toMatchSnapshot();
	});

	it('should match the snapshot when the store prop popup is true', () => {
		wrapper = shallow(<Main popup={{}} />);
		expect(wrapper).toMatchSnapshot();
	});

	it('should have the appropriate default state', () => {
		expect(wrapper.state()).toEqual({
			query: '',
			results: null
		});
	});

	it('should update state when handleChange is invoked and the input has a value', () => {
		expect(wrapper.state('query')).toEqual('');

		wrapper.instance().handleChange(mockEvent);

		expect(wrapper.state('query')).toEqual('beef');
	});

	it('should update query state when handleChange is invoked and the event target has no value', () => {
		const mockEmptyEvent = { target: { value: '' } };

		wrapper.instance().handleChange(mockEvent);

		expect(wrapper.state('query')).toEqual('beef');

		wrapper.instance().handleChange(mockEmptyEvent);

		expect(wrapper.state('query')).toEqual('');
	});

	it('should call searchQuests when handleSubmit is invoked', () => {
		wrapper.instance().handleSubmit(mockEvent);

		expect(helper.searchQuests).toHaveBeenCalled();
	});

	it('should update results state with the results if they exist when handleSubmit is invoked', () => {
		expect(wrapper.state('results')).toEqual(null);

		wrapper.instance().handleSubmit(mockEvent);

		expect(wrapper.state('results')).toEqual(mockQuests);
	});

	it('should update results state to null when handleChange is invoked and there is no event target value', () => {
		const mockEmptyEvent = { target: { value: '' } };

		wrapper.instance().handleSubmit(mockEvent);

		expect(wrapper.state('results')).toEqual(mockQuests);

		wrapper.instance().handleChange(mockEmptyEvent);

		expect(wrapper.state('results')).toEqual(null);
	});

	it('should update the results state to null when clearResults is invoked', () => {
		wrapper.instance().handleSubmit(mockEvent);

		expect(wrapper.state('results')).toEqual(mockQuests);

		wrapper.instance().clearResults();

		expect(wrapper.state('results')).toEqual(null);
	});

	describe('Main event listeners', () => {
		it('should call handleChange when the input value changes', () => {
      expect(wrapper.state('query')).toEqual('');

      wrapper.find('.search-input').simulate('change', mockEvent);

      expect(wrapper.state('query')).toEqual('beef');
    });
    
    it('should call handleSubmit when the input value changes', () => {
      wrapper.find('.search-bar').simulate('submit', mockEvent);

      expect(mockEvent.preventDefault).toHaveBeenCalled();
    });
  });
  
  describe('mapStateToProps', () => {
    it('should return an object with popup and quests data', () => {
      const expected = { popup: false, quests: mockQuests };

      const mappedProps = mapStateToProps({ popup: false, quests: mockQuests });

      expect(mappedProps).toEqual(expected);
    });
  })

  describe('mapDispatchToProps', () => {
    it('calls dispatch with an addQuests action when login is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = addQuest(mockQuest);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.addQuest(mockQuest);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});
