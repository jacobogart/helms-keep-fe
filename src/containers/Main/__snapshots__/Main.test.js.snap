// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should match the snapshot when the store prop popup is false 1`] = `
<main
  className="Main"
>
  <div
    className="main-header"
  >
    <h1
      className="page-title"
    >
      My Quests
    </h1>
    <div
      className="controls"
    >
      <form
        className="search-bar"
        onSubmit={[Function]}
      >
        <input
          className="search-input"
          onChange={[Function]}
          placeholder="Search your quests..."
        />
        <button
          className="search-btn-container"
          type="submit"
        >
          <i
            className="fas fa-search"
            role="button"
          />
        </button>
      </form>
    </div>
  </div>
  <ConnectFunction
    results={null}
  />
  <NavLink
    className="add-quest-btn"
    to="/new-quest"
    type="button"
  >
    <i
      className="fas fa-plus"
    />
  </NavLink>
</main>
`;

exports[`Main should match the snapshot when the store prop popup is true 1`] = `
<main
  className="Main"
>
  <div
    className="main-header"
  >
    <h1
      className="page-title"
    >
      My Quests
    </h1>
    <div
      className="controls"
    >
      <form
        className="search-bar"
        onSubmit={[Function]}
      >
        <input
          className="search-input"
          onChange={[Function]}
          placeholder="Search your quests..."
        />
        <button
          className="search-btn-container"
          type="submit"
        >
          <i
            className="fas fa-search"
            role="button"
          />
        </button>
      </form>
    </div>
  </div>
  <ConnectFunction
    results={null}
  />
  <NavLink
    className="add-quest-btn"
    to="/new-quest"
    type="button"
  >
    <i
      className="fas fa-plus"
    />
  </NavLink>
</main>
`;
